x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/rules_tables.tsx

```diff
@@ -219,6 +219,12 @@ export const RulesTables = React.memo<RulesTableProps>(
            */
           if (isSelectAllCalled.current) {
             isSelectAllCalled.current = false;
+            // Handle special case of unselecting all rules via checkbox
+            // after all rules were selected via Bulk select.
+            if (selected.length === 0) {
+              setIsAllSelected(false);
+              setSelectedRuleIds([]);
+            }
```

Comment:
This logical branch gets executed only after users click the "Select all X rules", so it probably wouldn't help when users change table selection by clicking on checkboxes.

Reply:
We've synced with @jpdjere on the remaining rule selection edge case and discussed that it would be better to fix the rule selection synchronization problem on the EUI level. Here's a ticket for that: https://github.com/elastic/eui/issues/6184. 

I think it is okay if we fix it in a follow-up PR.